pinot-core/src/main/java/com/linkedin/pinot/core/data/recordtransformer/DataTypeTransformer.java

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.data.recordtransformer;
+
+import com.linkedin.pinot.common.data.FieldSpec;
+import com.linkedin.pinot.common.data.Schema;
+import com.linkedin.pinot.common.utils.StringUtil;
+import com.linkedin.pinot.common.utils.primitive.ByteArray;
+import com.linkedin.pinot.core.data.GenericRow;
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The {@code DataTypeTransformer} class will convert the values to follow the data types in {@link FieldSpec}.
+ */
+public class DataTypeTransformer implements RecordTransformer {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataTypeTransformer.class);
+
+  private static final Map<Class, PinotDataType> SINGLE_VALUE_TYPE_MAP = new HashMap<>();
+  private static final Map<Class, PinotDataType> MULTI_VALUE_TYPE_MAP = new HashMap<>();
+
+  static {
+    SINGLE_VALUE_TYPE_MAP.put(Boolean.class, PinotDataType.BOOLEAN);
+    SINGLE_VALUE_TYPE_MAP.put(Byte.class, PinotDataType.BYTE);
+    SINGLE_VALUE_TYPE_MAP.put(Character.class, PinotDataType.CHARACTER);
+    SINGLE_VALUE_TYPE_MAP.put(Short.class, PinotDataType.SHORT);
+    SINGLE_VALUE_TYPE_MAP.put(Integer.class, PinotDataType.INTEGER);
+    SINGLE_VALUE_TYPE_MAP.put(Long.class, PinotDataType.LONG);
+    SINGLE_VALUE_TYPE_MAP.put(Float.class, PinotDataType.FLOAT);
+    SINGLE_VALUE_TYPE_MAP.put(Double.class, PinotDataType.DOUBLE);
+    SINGLE_VALUE_TYPE_MAP.put(String.class, PinotDataType.STRING);
+    SINGLE_VALUE_TYPE_MAP.put(byte[].class, PinotDataType.BYTES);
+    SINGLE_VALUE_TYPE_MAP.put(ByteArray.class, PinotDataType.BYTES);
+
+    MULTI_VALUE_TYPE_MAP.put(Byte.class, PinotDataType.BYTE_ARRAY);
+    MULTI_VALUE_TYPE_MAP.put(Character.class, PinotDataType.CHARACTER_ARRAY);
+    MULTI_VALUE_TYPE_MAP.put(Short.class, PinotDataType.SHORT_ARRAY);
+    MULTI_VALUE_TYPE_MAP.put(Integer.class, PinotDataType.INTEGER_ARRAY);
+    MULTI_VALUE_TYPE_MAP.put(Long.class, PinotDataType.LONG_ARRAY);
+    MULTI_VALUE_TYPE_MAP.put(Float.class, PinotDataType.FLOAT_ARRAY);
+    MULTI_VALUE_TYPE_MAP.put(Double.class, PinotDataType.DOUBLE_ARRAY);
+    MULTI_VALUE_TYPE_MAP.put(String.class, PinotDataType.STRING_ARRAY);
+  }
+
+  private final Map<String, PinotDataType> _dataTypes = new HashMap<>();
+
+  public DataTypeTransformer(Schema schema) {
+    for (Map.Entry<String, FieldSpec> entry : schema.getFieldSpecMap().entrySet()) {
+      _dataTypes.put(entry.getKey(), PinotDataType.getPinotDataType(entry.getValue()));
+    }
+  }
+
+  @Override
+  public GenericRow transform(GenericRow record) {
+    for (Map.Entry<String, PinotDataType> entry : _dataTypes.entrySet()) {
+      String column = entry.getKey();
+      PinotDataType dest = entry.getValue();
+      Object value = record.getValue(column);
+
```

Comment:
should you continue if it is a time field-spec?
Somehow i feel it is better to fold in time transformation into record transformation (like before).

Reply:
By this point, all fields should be treated the same (such as expression fields). DataTypeTransformer should be the last transformer which guarantees that all the values follow the field spec.